### DISCUSSION TITLE
[Question] Calculation Frequency of "attitude_pid_controller.c"

### DISCUSSION THREAD

**Discussion by doltae09** [bitcraze/discussions#670] [Q&A] [answered]
Hello! I want to know the calculation frequency of "attitude_pid_controller.c."

I first noticed that in the following guidance, "attitude_pid_controller.c" is calculated as 500 Hz.
https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/

But I wrote the code as below to see if this is.
![image](https://user-images.githubusercontent.com/79627957/233102201-88a8e565-2139-4974-b0e2-183c3bf81f0f.png)

As a result, it is confirmed in the Client as shown in the following figure(sampling time of parameter in Client= 10ms).
![image](https://user-images.githubusercontent.com/79627957/233103821-6b8a01c0-3c09-4fdc-b55d-1bfc67306a56.png)

If you look at the unit of "usertime()", it's [micro-second] by annotation.
![image](https://user-images.githubusercontent.com/79627957/233105773-fd14ccc9-e6f4-4125-abce-06446de361dd.png)

This is 1 [micro-second] = 0.000001 [s], so the calculation frequency of "attitude_pid_controller.c" is determined to be 0.000008 [s]. 
Here I'm confused about the concept of the calculation frequency. I wonder if I reviewed it correctly.

**Comment by krichardsson** [comment, marked as answer]
Hi!

It looks like your code is measuring the time spent in the two functions `attitudeControllerCorrectRatePID()` and `attitudeControllerCoreectAttitudePID()`. The rate mentioned in the documentation is related to the number of calls / second rather than time it takes to execute the function.

The rate mechanism is implemented using the [RATE_DO_EXECUTE](https://github.com/bitcraze/crazyflie-firmware/blob/9956e68e10ba52855634c08abde62fb093707814/src/modules/interface/stabilizer_types.h#L372) macro. 

In your case you can find the code in the [`controller_pid.c`](https://github.com/bitcraze/crazyflie-firmware/blob/9956e68e10ba52855634c08abde62fb093707814/src/modules/src/controller/controller_pid.c#L97-L109).

```C
  if (RATE_DO_EXECUTE(ATTITUDE_RATE, stabilizerStep)) {
    // ...
    attitudeControllerCorrectAttitudePID(state->attitude.roll, state->attitude.pitch, state->attitude.yaw,
                                attitudeDesired.roll, attitudeDesired.pitch, attitudeDesired.yaw,
                                &rateDesired.roll, &rateDesired.pitch, &rateDesired.yaw);
```

where `ATTITUDE_RATE` is defined as 500Hz.
The stabilizer loop runs at 1000 Hz and the `RATE_DO_EXECUTE` macro will essentially return `true` every second time in this case.

Tip: If you want to measure rates, you can also use the `STATS_CNT_RATE_EVENT` macro.

**Reply by doltae09** [reply]
I got it! Thank you, sir! I can understand much more about crazyflie-firmware!
